### PR TITLE
Add some changes to make connection recovery easier for senders

### PR DESCRIPTION
**Pull request by richardpark-msft**
As part of an investigation into an issue in Event Hubs I found a couple of cases that were problematic when attempting to do recovery.

First, if an error happens (and the sender is subsequently closed) that same error will get repeatedly returned by the sender. The error itself is actually a throttling error (and was relevant the first time it was returned) but after that we really should be indicating the link will _never_ work again because it has been closed.

The second is just exposing the link ID so it's simpler, from the outside, to tell one sender apart from another. As part of a separate change I want to make in Event Hubs I need to be able to see if the sender that I'm about to close/remove is actually the same one I started with (if it's not then I don't have to recover since the 'bad' link has already been cycled out).

### CODE REVIEW

File: sender.go [hunk 1]
```diff
@@ -16,6 +16,11 @@ type Sender struct {
 	nextDeliveryTag uint64
 }
 
+// Id() is the ID of the link used for this Sender.
+func (s *Sender) Id() string {
```
Comment (jhendrixMSFT) [inline sender.go:20]: Nit: acronyms should be all upper-case.

File: sender_test.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+func TestClosedSenderReturnsErrClosed(t *testing.T) {
+	// this feels a bit _too_ fake, should revisit.
+	link, err := newLink(newSession(nil, 0), &Receiver{}, nil)
+	require.NoError(t, err)
+
+	sender := &Sender{link: link}
+
+	// simulate a detach happening before the send()
+	// link.muxDetach()
+	close(link.done)
```
Comment (jhendrixMSFT) [inline sender_test.go:19]: Can we not call `sender.Close()` instead?

File: sender.go (hunk 1 above)
Comment (richardpark-msft) [reply sender.go:20]: That's the kind of thing I'm on this team for :)

File: sender.go [hunk 3]
```diff
@@ -26,6 +31,18 @@ type Sender struct {
 // additional messages can be sent while the current goroutine is waiting
 // for the confirmation.
 func (s *Sender) Send(ctx context.Context, msg *Message) error {
+	select {
+	case <-s.link.detached:
```
Comment (jhendrixMSFT) [inline sender.go:35]: Do we need a similar check on the `Receiver` to return `ErrLinkClosed`?

File: sender_test.go (hunk 2 above)
Comment (richardpark-msft) [reply sender_test.go]: This brings up an interesting question that I didn't realize until you asked this.

TLDR: I think this code is 'correct', but only because of the way the code is _currently_.

Longer:

There are (apparently) two channels that indicate a closed state with the `link`. `done` and `close`. 

* `close(l.done)` gets called if the service requests you close (ie: detach) or if the server sends back an error, in which case we do the detach. I renamed it to `detached` just now because that's what it's meant to represent. 
* `close(l.close)` gets called if a caller calls l.Close() explicitly. So it represents a user-initiated close.

I think there's value in keeping these separate (a user calling Close() means the link should not reopen, whereas the service asking you to detach does NOT (ie, detach should be considered recoverable)).

What we could do is actually create an ErrLinkDetached or similar, that lets you distinguish these programatically. That way our recovery code can still take advantage of this, but also do the right thing if the link was closed (ie - don't reopen it!).

File: sender.go (hunk 3 above)
Comment (richardpark-msft) [reply sender.go:35]: I just updated in the other comment - I think we should do a similar thing here but perhaps have a new error code (ErrLinkDetached) so we can distinguish between explicit close by caller, and close from service (ie: detach)

File: sender_test.go (hunk 2 above)
Comment (richardpark-msft) [reply sender_test.go]: Just FYI: I added a function (link.Check) that returns either ErrLinkClosed or ErrLinkDetached when called and migrated my code into there. I think it's doing the right thing in a better way (check that the link is valid) and we can expand it as we find more cases (although I don't anticipate any).

File: sender.go (hunk 3 above)
Comment (richardpark-msft) [reply sender.go]: Receiver seems slightly more complicated - I think we should design this a bit better with Receiver in mind in a future PR.

File: sender.go [hunk 4]
```diff
@@ -26,6 +31,12 @@ type Sender struct {
 // additional messages can be sent while the current goroutine is waiting
 // for the confirmation.
 func (s *Sender) Send(ctx context.Context, msg *Message) error {
+	err := s.link.Check()
```
Comment (jhendrixMSFT) [inline sender.go:34]: Nit: condense into the `if` block as `err` is immediately returned.
```go
if err := s.link.Check(); err != nil {
    return err
}
```

**Review by jhendrixMSFT — APPROVED** [review]

File: sender.go (hunk 4 above)
Comment (richardpark-msft) [reply sender.go]: Done!

File: link.go [hunk 5]
```diff
@@ -638,6 +646,19 @@ func (l *link) muxHandleFrame(fr frameBody) error {
 	return nil
 }
 
+// Check checks the link state, returning an error if the link is closed (ErrLinkClosed) or if
+// it is in a detached state (ErrLinkDetached)
+func (l *link) Check() error {
+	select {
+	case <-l.detached:
+		return ErrLinkDetached
```
Comment (serbrech) [inline link.go:654]: any case where both channel have a value?
if both have an error in it, the returned value will be undefined I think? (either or...)
but, probably does not matter since in both case, the link is not good.

File: sender.go [hunk 6]
```diff
@@ -26,6 +31,10 @@ type Sender struct {
 // additional messages can be sent while the current goroutine is waiting
 // for the confirmation.
 func (s *Sender) Send(ctx context.Context, msg *Message) error {
+	if err := s.link.Check(); err != nil {
```
Comment (serbrech) [inline sender.go:34]: so from a consumer point of view, the Check() func allows you to check periodically for the state of the link, but not to be notified of the link going bust unfortunately.

Comment (richardpark-msft) [reply sender.go:34]: Right, just a simple way to standardize on "is this link even good".

The other option (which I considered) was to provide "IsDetached" and "IsClosed" exports, but it seemed simpler to just take advantage of the predefined errors. 

In rhea (JS AMQP), for instance, they have these exported as 'is closed' and 'is remote closed' variables (well, and a specialized 'onDetached' handler).

> but not to be notified of the link going bust unfortunately.

For senders this isn't usually too bad (since you're basically always initiating an operation). I think your question is more about receivers. You (and I and @jhendrixMSFT) will definitely need to give that some thought!

File: link.go (hunk 5 above)
Comment (richardpark-msft) [reply link.go:654]: I had a similar conversation with @jhendrixMSFT - this is probably something we should discuss more (I think this PR is fine, but it'd be good to think about how we represent it).

For instance, if you received an error and then called 'Check' you'd get back detached. Reacting to that by recreating a link feels right (ie: detached is a retryable error at the link level). 

However, if you were detached and then also called Close() - that's no longer retryable. You should consider that someone is trying to shut down. So I'd like Close() to take precedence over Detach in that case.

Given that `select+case` statements are pseudo-random, is there a Go way of doing that? I could check to see (in the l.detached case) if l.close is also closed. Would that be the right way of doing it?

File: sender.go (hunk 6 above)
Comment (serbrech) [reply sender.go:34]: I was thinking of exposing an error channel, so that from the listener point of view, you can be notified as soon as this happens.

For example, for a sender that sends very few messages, say 1 message every few minutes. I'd want to know that the link died before I send, so that I can reconnect as soon as possible, and be healthy when the time comes to send a message.

File: link.go (hunk 5 above)
Comment (serbrech) [reply link.go:654]: Right i think that's a minor issue.
Your implementation would work.

I do think that a channel would allow the errors to be surfaced as they come, and so the consumer can handle them right away, and not get one or the other.
